src/ray/raylet/scheduling/scheduler_resource_reporter.cc

```diff
@@ -64,103 +67,66 @@ void SchedulerResourceReporter::FillResourceUsage(
-
-    for (const auto &resource : resources) {
-      // Add to `resource_loads`.
-      const auto &label = resource.first;
-      const auto &quantity = resource.second;
-      (*resource_loads)[label] += quantity * count;
-
-      // Add to `resource_load_by_shape`.
-      (*by_shape_entry->mutable_shape())[label] = quantity;
-    }
-    int num_ready = by_shape_entry->num_ready_requests_queued();
-    by_shape_entry->set_num_ready_requests_queued(num_ready + count);
-    by_shape_entry->set_backlog_size(TotalBacklogSize(scheduling_class));
-  }
-  for (const auto &pair : infeasible_tasks_) {
-    const auto &scheduling_class = pair.first;
-    if (num_reported++ >= max_resource_shapes_per_load_report_ &&
-        max_resource_shapes_per_load_report_ >= 0) {
-      // TODO (Alex): It's possible that we skip a different scheduling key which contains
-      // the same resources.
-      skipped_requests++;
-      break;
-    }
-    const auto &resources =
-        TaskSpecification::GetSchedulingClassDescriptor(scheduling_class)
-            .resource_set.GetResourceMap();
-    const auto &queue = pair.second;
-    const auto &count = queue.size();
-
-    auto by_shape_entry = resource_load_by_shape->Add();
-    for (const auto &resource : resources) {
-      // Add to `resource_loads`.
-      const auto &label = resource.first;
-      const auto &quantity = resource.second;
-      (*resource_loads)[label] += quantity * count;
-
-      // Add to `resource_load_by_shape`.
-      (*by_shape_entry->mutable_shape())[label] = quantity;
-    }
+      if (visited.count(scheduling_class) != 0) {
+        continue;
+      }
 
-    // If a task is not feasible on the local node it will not be feasible on any other
-    // node in the cluster. See the scheduling policy defined by
-    // ClusterResourceScheduler::GetBestSchedulableNode for more details.
-    int num_infeasible = by_shape_entry->num_infeasible_requests_queued();
-    by_shape_entry->set_num_infeasible_requests_queued(num_infeasible + count);
-    by_shape_entry->set_backlog_size(TotalBacklogSize(scheduling_class));
-  }
+      const auto &resources =
+          TaskSpecification::GetSchedulingClassDescriptor(scheduling_class)
+              .resource_set.GetResourceMap();
+      auto by_shape_entry = resource_load_by_shape->Add();
+
+      if (count != 0) {
+        for (const auto &resource : resources) {
+          // Add to `resource_loads`.
+          const auto &label = resource.first;
+          const auto &quantity = resource.second;
+          (*resource_loads)[label] += quantity * count;
+
+          // Add to `resource_load_by_shape`.
+          (*by_shape_entry->mutable_shape())[label] = quantity;
+        }
+      }
+
+      if (is_infeasible) {
+        by_shape_entry->set_num_infeasible_requests_queued(count);
+      } else {
+        by_shape_entry->set_num_ready_requests_queued(count);
+      }
+      by_shape_entry->set_backlog_size(TotalBacklogSize(scheduling_class));
```

Comment:
Now I'm wondering if the existing code is wrong: for the same scheduling_class, we may create multiple by_shape_entry since the same scheduling_class can appear in multiple task queues, and each sets backlog size to the same. Won't autoscaler get wrong result if it sums them up?

Reply:
I think it works out because the `resource_load_by_shape` will just look like `[{shape={"CPU":1}, num_queued=2}, {shape={"CPU":1}, num_queued=3}]` And the autoscaler won't care and will just expand it out to be 5 * {CPU: 1} demands.

Reply:
Yea, `num_queued` is fine since autoscaler should just add them up. I'm worried about `backlog_size` since it sets to the same total for same shape: `[{shape={"CPU":1}, num_queued=2, backlog_size=100}, {shape={"CPU":1}, num_queued=3, backlog_size=100}]`, total backlog_size should be 100 not 200? I need to read autoscaler code to see how backlog_size is used.

Reply:
oh you're right that's a bug in the existing code (assuming we're parallelizing lease requests, which tbh i don't remember if that's on by default 😅)

Reply:
Yea it's on by default with parallelism 10 (even if it's off we still have the same issue I think since we can still have multiple by_shape_entry with the same shape caused by lease requests from different workers).